nexus/db-queries/src/db/queries/instance.rs

```diff
@@ -0,0 +1,258 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Implement a query for updating an instance and VMM in a single CTE.
+
+use async_bb8_diesel::AsyncRunQueryDsl;
+use diesel::prelude::QueryResult;
+use diesel::query_builder::{Query, QueryFragment, QueryId};
+use diesel::sql_types::{Nullable, Uuid as SqlUuid};
+use diesel::{pg::Pg, query_builder::AstPass};
+use diesel::{Column, ExpressionMethods, QueryDsl, RunQueryDsl};
+use nexus_db_model::{
+    schema::{instance::dsl as instance_dsl, vmm::dsl as vmm_dsl},
+    InstanceRuntimeState, VmmRuntimeState,
+};
+use uuid::Uuid;
+
+use crate::db::pool::DbConnection;
+use crate::db::update_and_check::UpdateStatus;
+
+/// A CTE that checks and updates the instance and VMM tables in a single
+/// atomic operation.
+//
+// The single-table update-and-check CTE has the following form:
+//
+// WITH found   AS (SELECT <primary key> FROM T WHERE <primary key = value>)
+//      updated AS (UPDATE T SET <constraints> RETURNING *)
+// SELECT
+//      found.<primary key>
+//      updated.<primary key>
+//      found.*
+// FROM
+//      found
+// LEFT JOIN
+//      updated
+// ON
+//      found.<primary_key> = updated.<primary_key>;
+//
+// The idea behind this query is to have separate "found" and "updated"
+// subqueries for the instance and VMM tables, then use those to create two more
+// subqueries that perform the joins and yield the results, along the following
+// lines:
+//
+// WITH vmm_found AS (SELECT(SELECT id FROM vmm WHERE vmm.id = id) AS id),
```

Comment:
Honestly, the hardest part was figuring out that I could do the `SELECT(SELECT(` trick to avoid having the query fail if one of the relevant objects was garbage-collected...

I'll give RFD 436 a look and see if I can leave a helpful comment about this.